File: 剑指offer/38.字符串的排列.py
from typing import List


class Solution:
    def permutation(self, s: str) -> List[str]:
        chars = list(s)

        path = []
        ans = []

        def help(index):
            if len(path) == len(s):
                ans.append(''.join(path))

            mem = set()
            for i in range(index, len(s)):
                if chars[i] in mem:
                    continue
                mem.add(chars[i])

                chars[i], chars[index] = chars[index], chars[i]
                path.append(chars[index])
                help(index + 1)
                chars[i], chars[index] = chars[index], chars[i]
                path.pop(-1)

        help(0)
        return ans

File: 剑指offer/38.字符串的排列.cpp
#include <algorithm>
#include <set>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<string> ans;
    int n = 0;

    void help(vector<char> candiate, string temp, vector<bool> visited)
    {
        if (temp.size() == candiate.size()) {
            ans.push_back(temp);
            return;
        }

        set<char> visited_c;
        for (int i = 0; i < candiate.size(); i++) {
            if (visited[i] or visited_c.count(candiate[i])) {
                continue;
            }

            visited_c.insert(candiate[i]);
            visited[i] = true;
            help(candiate, temp + candiate[i], visited);
            visited[i] = false;
        }
    }

    vector<string> permutation(string s)
    {
        vector<char> c_list;
        vector<bool> visited(s.size(), false);
        string temp = "";

        for (char c : s) {
            c_list.push_back(c);
        }

        help(c_list, temp, visited);
        return ans;
    }
};